### DISCUSSION TITLE
Extract headers using opentelemetry-http HeaderExtractor

### DISCUSSION THREAD

**Discussion by robin-thoene** [tokio-rs/axum#2657] [Q&A] [answered]
Hey you all,

I currently try to build an open telemetry tracing middleware.

To extract the parent context for spans I am following this example:

https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/tracing-http-propagator/src/server.rs

There you can see how the request.headers() are used to get the parent context.

The problem I have is that the request.headers() in my Axum handler are from type axum::http::HeaderMap, but the open telemetry extractor expects the type the http::HeaderMap

I am quite new to rust and I am pretty confused at the moment.

**Comment by jplatte** [comment, marked as answer]
Hey, it sounds like you are using two different versions of the `http` crate at the same time, and getting a type mismatch when trying to use `HeaderMap` from both interchangeably.

`axum::http` is simply a re-export of the `http` crate as you can see [here](https://docs.rs/axum/latest/axum/#reexports). As of axum 0.7, that's version 1.x.y of the `http` crate, before it was version 0.2.x.

**Comment by robin-thoene** [comment]
Hey, thanks for the response. I just checked my Cargo.lock. There are indeed two different versions of the `http` crate resolved. One of them has the version 0.2. I just looked up the dependencies in the opentelemetry project and there is the version 0.2 defined as workspace dependency.

I will try to investigate how to resolve the version conflict.

**Reply by jplatte** [reply]
Upgrading the http crate in the otel crates is unfortunately blocked: https://github.com/open-telemetry/opentelemetry-rust/issues/1427. But it should be possible to write some conversion code for the http 0.2 / 1.0 types. I'm not aware of an existing crate for that though.
